ntoskrnl/se/debug.c

```diff
@@ -0,0 +1,388 @@
+        DbgPrint("ACE mask right -> 0x%08lx\n", Ace->AccessMask);
+
+        Sid = SepGetSidFromAce(Ace->Header.AceType, Ace);
+        ASSERT(Sid);
+        RtlConvertSidToUnicodeString(&SidString, Sid, TRUE);
+        DbgPrint("ACE SID -> %wZ\n", &SidString);
+        RtlFreeUnicodeString(&SidString);
+
+        DbgPrint("ACE size -> %u\n", Ace->Header.AceSize);
+        DbgPrint("ACE flags:\n");
+        SepDumpAceFlags(Ace->Header.AceFlags);
+
+        /* Increment the ACE number as we go further */
+        AceNumber++;
+    }
+}
+
+/**
+ * @brief
+ * Dumps debug info of an Access Control List (ACL).
+ */
+static
+VOID
+SepDumpAclInfo(
+    _In_ PACL Acl,
+    _In_ BOOLEAN IsSacl)
+{
+    /* Dump relevant info */
+    DbgPrint("================== %s DUMP INFO ==================\n", IsSacl ? "SACL" : "DACL");
+    DbgPrint("ACL Revision -> %u\n", Acl->AclRevision);
+    DbgPrint("ACL size -> %u\n", Acl->AclSize);
+    DbgPrint("ACE count of ACL -> %u\n", Acl->AceCount);
+
+    /* Dump all the ACEs present on this ACL */
+    SepDumpAces(Acl);
+}
+
+/**
+ * @brief
+ * Dumps control flags of a security descriptor to the debugger.
+ */
+static
+VOID
+SepDumpSdControlInfo(
+    _In_ SECURITY_DESCRIPTOR_CONTROL SdControl)
+{
+    /* Dump the control mask flags in the following order */
+    if (SdControl & SE_OWNER_DEFAULTED)
+    {
+        DbgPrint("SE_OWNER_DEFAULTED\n");
+    }
+
+    if (SdControl & SE_GROUP_DEFAULTED)
+    {
+        DbgPrint("SE_GROUP_DEFAULTED\n");
+    }
+
+    if (SdControl & SE_DACL_PRESENT)
+    {
+        DbgPrint("SE_DACL_PRESENT\n");
+    }
+
+    if (SdControl & SE_DACL_DEFAULTED)
+    {
+        DbgPrint("SE_DACL_DEFAULTED\n");
+    }
+
+    if (SdControl & SE_SACL_PRESENT)
+    {
+        DbgPrint("SE_SACL_PRESENT\n");
+    }
+
+    if (SdControl & SE_SACL_DEFAULTED)
+    {
+        DbgPrint("SE_SACL_DEFAULTED\n");
+    }
+
+    if (SdControl & SE_DACL_UNTRUSTED)
+    {
+        DbgPrint("SE_DACL_UNTRUSTED\n");
+    }
+
+    if (SdControl & SE_SERVER_SECURITY)
+    {
+        DbgPrint("SE_SERVER_SECURITY\n");
+    }
+
+    if (SdControl & SE_DACL_AUTO_INHERIT_REQ)
+    {
+        DbgPrint("SE_DACL_AUTO_INHERIT_REQ\n");
+    }
+
+    if (SdControl & SE_SACL_AUTO_INHERIT_REQ)
+    {
+        DbgPrint("SE_SACL_AUTO_INHERIT_REQ\n");
+    }
+
+    if (SdControl & SE_DACL_AUTO_INHERITED)
+    {
+        DbgPrint("SE_DACL_AUTO_INHERITED\n");
+    }
+
+    if (SdControl & SE_SACL_AUTO_INHERITED)
+    {
+        DbgPrint("SE_SACL_AUTO_INHERITED\n");
+    }
+
+    if (SdControl & SE_DACL_PROTECTED)
+    {
+        DbgPrint("SE_DACL_PROTECTED\n");
+    }
+
+    if (SdControl & SE_SACL_PROTECTED)
+    {
+        DbgPrint("SE_SACL_PROTECTED\n");
+    }
+
+    if (SdControl & SE_RM_CONTROL_VALID)
+    {
+        DbgPrint("SE_RM_CONTROL_VALID\n");
+    }
+
+    if (SdControl & SE_SELF_RELATIVE)
+    {
+        DbgPrint("SE_SELF_RELATIVE\n");
+    }
```

Comment:
```suggestion
#define SEPRINT(x)    \
    if (SdControl & x) \
    { \
        DbgPrint(#x "\n"); \
    }

    SEPRINT(SE_OWNER_DEFAULTED);
    SEPRINT(SE_GROUP_DEFAULTED);
    SEPRINT(SE_DACL_PRESENT);
    SEPRINT(SE_DACL_DEFAULTED);
    SEPRINT(SE_SACL_PRESENT);
    SEPRINT(SE_SACL_DEFAULTED);
    SEPRINT(SE_DACL_UNTRUSTED);
    SEPRINT(SE_SERVER_SECURITY);
    SEPRINT(SE_DACL_AUTO_INHERIT_REQ);
    SEPRINT(SE_SACL_AUTO_INHERIT_REQ);
    SEPRINT(SE_DACL_AUTO_INHERITED);
    SEPRINT(SE_SACL_AUTO_INHERITED);
    SEPRINT(SE_DACL_PROTECTED);
    SEPRINT(SE_SACL_PROTECTED);
    SEPRINT(SE_RM_CONTROL_VALID);
    SEPRINT(SE_SELF_RELATIVE);
#undef SEPRINT
```
(same, I didn't tested it)